core/src/main/java/org/apache/iceberg/TableUtil.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class TableUtil {
+  private TableUtil() {}
+
+  /** Returns the format version of the given table */
+  public static int formatVersion(Table table) {
+    Preconditions.checkArgument(null != table, "Invalid table: null");
+
+    // SerializableMetadataTable is a subclass of SerializableTable but does not support
+    // operations()
+    if (table instanceof HasTableOperations
```

Comment:
I think we still have a problem that getting `SerializableTable` here would require loading the metadata location and sending a request to the storage. I would consider capturing format version as a field in `SerializableTable` and throwing `UnsupportedOperationException` in `SerializableMetadataTable` that implements it. Then the whole logic can be structured as follows:

```
if (table instanceof SerializableTable) {
  SerializableTable serializableTable = (SerializableTable) table;
  return serializableTable.formatVersion();
} else if (table instanceof HasTableOperations) {
  TableOperations ops = ((HasTableOperations) table).operations();
  return ops.current().formatVersion();
} else {
  throw new IllegalArgumentException(
      String.format("%s does not have a format version", table.getClass().getSimpleName()));
}
```

What do you think, @nastra?

Reply:
> I think we still have a problem that getting SerializableTable here would require loading the metadata location and sending a request to the storage

Yes this is a trade-off here because I was trying to avoid having effectively a mix of a Util class and adding `formatVersion()` to the `SerializableTable` API. I don't have a strong opinion here so I've done this in https://github.com/apache/iceberg/pull/11620/commits/1d8ef0554c8361dc8eadec66e9a68fb27bf58557 to show its implications